### PR TITLE
fix(update): re-check transitive dependencies' own semver ranges during apm update

### PR DESCRIPTION
**Pull request by nadav-y**
## Summary

`apm update` was not picking up new versions of a **transitive** dependency, even when the intermediate package's manifest still allowed it:

```
org/pkg1 -> org/pkg2 (^1.0.0) -> org/pkg3 (^1.0.0)
```

Publishing a new `org/pkg3` matching `^1.0.0` was silently ignored by `apm update` run from `org/pkg1`, even though `org/pkg2`'s manifest hadn't changed.

## Root cause

`APMDependencyResolver._try_load_dependency_package` only calls `download_callback` (the entry point that re-resolves a semver range against the remote) when the install path doesn't already exist:

```python
if not install_path.exists():
    ...call download_callback...
```

`download_callback` itself (`resolve.py`) already contains a `_force_semver_resolve` fallthrough for exactly this case -- added by a prior fix ("Bug 1 fix on #1496") that cites the same symptom -- but it's dead code: the resolver's own gate above is the *only* thing that decides whether `download_callback` runs at all, and that prior fix only forces the path through for **direct** dependencies, via a pre-purge pass (`_purge_cached_semver_paths_for_update`) that deletes the install path before resolution starts. A transitive dependency's install path can't be pre-purged the same way -- its existence isn't knowable until its parent's manifest is read, which happens *during* resolution, not before it. So a transitive dependency's own semver range is never re-evaluated, no matter how many newer matching versions have been published.

## Fix

Widen the resolver's own gate with `APMDependencyResolver._should_force_recheck(dep_ref)` -- the identical predicate already used by `download_callback`'s `_force_semver_resolve` -- gated on a new `update_refs` constructor flag threaded from `resolve.py`:

```python
if not install_path.exists() or self._should_force_recheck(dep_ref):
```

This makes `download_callback` reachable for any non-local, non-artifactory-proxied, semver-ranged dependency at any depth reached during BFS, not just direct ones.

That's the entire fix. No backup/rollback code is added: `download_callback` already calls `staging_session.prepare_path(install_path)` (an atomic move into the resolution's rollback-scoped staging directory, from #2155's `ResolutionStagingSession`) immediately before overwriting, generically for whatever reaches that point in the callback. That already covers the newly-reachable transitive case with zero additional code -- verified live below.

**Scope note:** an earlier version of this PR was stacked on #2050 and also carried a staging/rollback mechanism (`update_backup.py`) plus a fail-closed fix for it. #2050 has since been closed as superseded by the already-merged #2155, which replaced that whole mechanism with the generalized `InstallTransaction`/`ResolutionStagingSession`. This PR has been rebuilt from scratch on top of current `main` (which already includes #2155) and now contains **only** the transitive-recheck fix described above -- confirmed the new architecture doesn't have the exception-swallowing flaw a review panel found in the old `update_backup.py` design (traced the full rollback chain: `ResolutionStagingSession.rollback()` -> `safe_rmtree()` -> `robust_rmtree(..., ignore_errors=False)` raises by default, nothing catches it), so no equivalent fix is needed here.

## Test plan

- [x] New unit tests in `test_apm_resolver_edge_cases.py` (`TestShouldForceRecheck`, `TestTryLoadDependencyPackageForceRecheck`) covering the recheck predicate and the widened gate across semver/literal ref kinds, local deps, artifactory-proxied deps, and `update_refs` on/off.
- [x] Full unit suite: 18403 passed, 7 pre-existing unrelated failures (`mcp_integrator`/`global_mcp_scope`/`test_helpers`/`lifecycle_executor`), 0 new failures.
- [x] Live repro against a real registry (`chain2/pkg1 -> chain2/pkg2 -> chain/pkg3`, all `^1.0.0`):
  - Fresh install, then successive `chain/pkg3` publishes + `apm update`: transitive update correctly applies each time.
  - Out-of-range major bump (e.g. `3.0.0` against `^1.0.0`): correctly ignored.
  - Non-interactive `apm update` (no TTY, no `--yes`) with a pending transitive update: aborts cleanly -- `apm.lock.yaml` hash unchanged, on-disk content untouched, no orphaned `.apm-resolution-staging/` directory.
  - Happy path (`-y`): applies cleanly, lockfile and content both updated.
  - Exact-pin-then-semver chain (`pkg1` pins `pkg2` at an exact version, `pkg2` ranges `pkg3` via `^1.0.0`): `pkg3` still updates through the exactly-pinned intermediate, confirming the recheck is per-node, not just per-chain.
  - **Confirmed #2155 alone does not fix this**: reverted just this PR's two source files against otherwise-unmodified current `main`, reproduced the exact same silent-ignore bug, then restored the fix and confirmed it resolves cleanly -- ruling out any chance the bug was already fixed elsewhere.
- [x] CI green on all 13 checks (Spec conformance gate, Lint, Build & Test both shards, CodeQL, etc.).

### CODE REVIEW

File: src/apm_cli/install/phases/update_backup.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+        # restore_update_backups can compute get_install_path() directly,
+        # without needing to look this dep up in all_apm_deps/
+        # deps_to_install afterward. Those two lists are direct-only /
+        # only-populated-on-success respectively -- a transitive dep
+        # backed up here and then caught by a resolution failure elsewhere
+        # (before deps_to_install is ever set) would otherwise be
+        # unresolvable and its backup permanently orphaned.
+        ctx.update_backups[dep_key] = (dep_ref, backup_path)
+        return True
+    return False
```
[the remark this answers is not in the document]
Comment (nadav-y) [reply src/apm_cli/install/phases/update_backup.py]: Good catch -- fixed in 813d1fbc. backup_before_overwrite no longer suppresses exceptions during staging; it now raises once a backup is actually required (plan_callback set + install_path exists), so the caller never proceeds to overwrite without a rollback point. The existing resolve-phase exception handler in pipeline.py (added for #2050's review) restores any backups already staged for other deps this run before re-raising, so the whole update aborts cleanly instead of silently losing content. Added a regression test simulating a rename failure and asserting it raises with install_path left untouched.

**Comment by sergio-sisternes-epam** [comment]
## APM Review Panel: `ship_with_followups`

> Depth-correct transitive semver re-resolution is structurally sound; ship with follow-up issues for path containment, decline-path test gap, and predicate dedup.

cc @nadav-y @danielmeppiel @sergio-sisternes-epam -- a fresh advisory pass is ready for your review.

This PR closes a real correctness gap in apm update: transitive dependencies with semver ranges were silently skipped during re-resolution because download_callback only fired when an install path didn't already exist. The fix is architecturally clean -- a new backup/restore module with fail-closed semantics, a predicate gate that short-circuits to zero cost on apm install, and 577 lines of new unit tests. All nine panelists agree the change is directionally correct and none raised a blocking finding. The strongest consensus signal is the duplicated predicate between _should_force_recheck and _force_semver_resolve, flagged independently by python-architect, devx-ux, and performance. The 'kept in sync deliberately' comment is honest but not mechanical enforcement; extracting a shared function is low-risk, high-payoff, and should land before the next contributor touches either site.

The supply-chain finding on path containment deserves strategic weight beyond its technical scope. APM's security model routes all path deletions through ensure_path_within / safe_rmtree chokepoints -- ad-hoc regex sanitization in a new module is a precedent violation, not just a style issue. The fix is zero-behavioral-change (one import, one call), and leaving it out creates a citable gap if APM's supply-chain posture is ever audited. I side with the supply-chain expert: this should be a fast follow-up issue, not deferred.

The test-coverage expert's two missing-evidence findings carry real weight. The decline-path restore assertion (outcome: missing on a secure-by-default surface) is the most important gap: the entire user promise of 'say no and nothing changes' has no automated guardrail proving restore_update_backups is called with keep_new=False. This is a regression trap waiting to happen. The integration-tier gap is lower urgency -- the cross-module contract is covered by unit mocks, and the backup/restore I/O is simple enough that unit coverage is defensible for now -- but it should be filed as a follow-up to close before the next release.

**Dissent.** No inter-panelist disagreements on severity classification. Three panelists (python-architect, devx-ux, performance) independently flagged the duplicated predicate; python-architect classified it recommended while devx-ux and performance classified it nit. CEO sides with python-architect's recommended classification: two copies of a gate predicate that must agree on a correctness-critical path is a maintenance hazard, not a style preference.

**Aligned with:** Pragmatic-as-npm: apm update now reaches full transitive depth for semver deps, matching the depth-correctness users expect from any mature package manager. Zero regression on apm install. | Secure-by-default: Fail-closed on backup-staging failure is correct; path containment gap is a fast follow-up. | OSS-community-driven: 577 lines of new tests from a contributor signal engineering rigor that builds long-term trust.

**Growth signal.** This PR is a trust-builder worth surfacing at release time. Frame as 'apm update is now depth-correct for semver across transitive dependencies' -- it positions APM alongside mature package managers and closes a silent-failure gap that would become a FAQ-generator as transitive dep trees grow.

### Panel summary

| Persona | B | R | N | Takeaway |
|---|---|---|---|---|
| Python Architect | 0 | 1 | 2 | Well-decomposed backup/restore module with correct fail-closed semantics; duplicated predicate is the main structural concern. |
| CLI Logging Expert | 0 | 2 | 1 | Backup/restore lifecycle is completely silent -- no verbose_detail on stage, no user-facing message on restore after decline/abort. |
| DevX UX Expert | 0 | 0 | 2 | Transitive semver re-resolution fix is UX-sound: update reaches full depth, declined plans restore cleanly, install is unaffected, no command-surface changes. |
| Supply Chain Security Expert | 0 | 1 | 1 | Backup module uses ad-hoc sanitization instead of the sanctioned path_security guards; suppress(Exception) in restore silences restore failures. |
| OSS Growth Hacker | 0 | 0 | 3 | High-trust correctness fix with excellent CHANGELOG prose; minor release-note framing opportunity missed but nothing blocks merge. |
| Auth Expert | 0 | 0 | 0 | No auth surface affected: auth_resolver still flows to resolver unchanged; new backup/restore module is auth-agnostic; wider download_callback reach reuses existing per-dep token resolution. |
| Doc Writer | 0 | 2 | 2 | Bug 3 (fail-closed staging failure) has no CHANGELOG entry; update.md omits transitive semver recheck behavior; two nits on mechanism wording and attribution style. |
| Test Coverage Expert | 0 | 2 | 1 | Strong unit coverage on predicate + backup/restore module; pipeline decline path missing restore assertion; no integration-tier test for transitive semver re-resolution flow. |
| Performance Expert | 0 | 1 | 4 | Correct approach; added cost is O(unique_repos) network RTTs for semver transitive deps during apm update, zero regression on apm install; backup I/O is ~1 rename per dep. |

> B = blocking-severity findings, R = recommended, N = nits.
> Counts are signal strength, not gates. The maintainer ships.

### Top 5 follow-ups

1. **[Test Coverage Expert]** Add unit assertion that decline path calls restore_update_backups(ctx, keep_new=False) -- Missing automated guardrail on a secure-by-default surface: the user promise 'decline an update and nothing changes' has no test proving restore is invoked.
2. **[Supply Chain Security Expert]** Route backup_path through ensure_path_within and use safe_rmtree instead of _rrm in restore -- APM's security model requires all path deletions to flow through sanctioned chokepoints. Zero-behavioral-change fix.
3. **[Python Architect]** Extract shared predicate from _should_force_recheck and _force_semver_resolve -- Three panelists independently flagged the duplicated gate condition. A shared function prevents silent drift on a correctness-critical predicate.
4. **[Doc Writer]** Add CHANGELOG entry for Bug 3 (fail-closed on backup-staging failure) and update update.md to mention transitive semver re-evaluation -- Bug 3 has no CHANGELOG bullet; docs page scopes re-resolution to direct deps only, which is now inaccurate.
5. **[CLI Logging Expert]** Emit verbose_detail on backup staging and user-facing summary on restore after decline/abort -- Replacement for the removed _purge log; restore after decline should confirm rollback happened.

### Architecture

```mermaid
classDiagram
    direction LR
    class APMDependencyResolver {
        <<Strategy>>
        +_update_refs bool
        +_download_callback DownloadCallback
        +_should_force_recheck(dep_ref) bool
        +_try_load_dependency_package(dep_ref, parent_chain, parent_pkg) APMPackage
    }
    class InstallContext {
        <<ValueObject>>
        +plan_callback Any
        +update_backups dict
        +callback_downloaded dict
        +apm_dir Path
        +apm_modules_dir Path
    }
    class DependencyReference {
        <<ValueObject>>
        +ref_kind str
        +is_local bool
        +get_unique_key() str
        +get_install_path(modules_dir) Path
    }
    class backup_before_overwrite {
        <<Pure Function>>
        +__call__(ctx, dep_ref, install_path) bool
    }
    class restore_update_backups {
        <<Pure Function>>
        +__call__(ctx, keep_new) None
    }
    class download_callback {
        <<Closure>>
        +_force_semver_resolve bool
    }
    class run_install_pipeline {
        <<IOBoundary>>
    }
    APMDependencyResolver ..> DependencyReference : checks ref_kind
    APMDependencyResolver ..> download_callback : invokes when gate opens
    download_callback ..> backup_before_overwrite : calls before overwrite
    download_callback ..> InstallContext : writes callback_downloaded
    backup_before_overwrite ..> InstallContext : writes update_backups
    backup_before_overwrite ..> DependencyReference : reads unique key
    run_install_pipeline ..> restore_update_backups : calls on resolve failure and plan gate
    restore_update_backups ..> InstallContext : reads update_backups and callback_downloaded
    classDef touched fill:#fff3b0,stroke:#d47600
    class APMDependencyResolver:::touched
    class backup_before_overwrite:::touched
    class restore_update_backups:::touched
    class InstallContext:::touched
    class download_callback:::touched
```

```mermaid
flowchart TD
    A["apm update CLI"] --> B["run_install_pipeline\npipeline.py"]
    B --> C["_run_phase resolve\npipeline.py"]
    C --> D["APMDependencyResolver.resolve\napm_resolver.py"]
    D --> E{"install_path.exists()\nAND NOT _should_force_recheck?"}
    E -->|"Yes: skip"| F["Load cached APMPackage\nfrom disk"]
    E -->|"No: gate open"| G["download_callback\nresolve.py"]
    G --> H{"_force_semver_resolve?"}
    H -->|"Yes"| I["[NET] git ls-remote"]
    H -->|"No"| J["Registry or git path"]
    I --> K["[FS] backup_before_overwrite\nupdate_backup.py"]
    J --> K
    K --> L{"plan_callback set\nAND path exists?"}
    L -->|"Yes"| M["[FS] install_path.rename backup_path"]
    L -->|"No"| N["No-op"]
    M --> O["ctx.update_backups[dep_key] = entry"]
    N --> P["[NET] download/clone to install_path"]
    O --> P
    P --> Q["Return to BFS"]
    C -->|"Exception"| R["[FS] restore_update_backups\nkeep_new=False"]
    R --> S["Re-raise"]
    B --> T["build_update_plan"]
    T --> U["plan_callback confirm gate"]
    U -->|"Declined / abort"| V["[FS] restore_update_backups\nkeep_new=False"]
    U -->|"Confirmed"| W["[FS] restore_update_backups\nkeep_new=True"]
    V --> X["Return empty InstallResult"]
    W --> Y["Continue to integrate phase"]
```

### Recommendation

This is a correctness fix for a real silent-failure gap in apm update, architecturally sound, with strong test coverage from a contributor. No panelist raised a blocking finding. The five follow-ups (decline-path test assertion, path containment, predicate dedup, CHANGELOG/docs gap, verbose logging) should be filed as issues and resolved before the next release, but none block merge of this PR.

---

<details>
<summary>Full per-persona findings</summary>

#### Python Architect

- **[recommended]** _should_force_recheck duplicates _force_semver_resolve -- extract shared predicate at `src/apm_cli/deps/apm_resolver.py:886`
  APMDependencyResolver._should_force_recheck and download_callback's _force_semver_resolve encode the same boolean condition with a 'kept in sync deliberately' comment. Two copies of a gate predicate that must agree is a maintenance hazard. Extract `is_semver_recheck_eligible(dep_ref, update_refs)` into a shared location.
  *Suggested:* Extract to `deps/semver_utils.py` or a DependencyReference property.

- **[nit]** backup_before_overwrite uses getattr defensively on a typed dataclass at `src/apm_cli/install/phases/update_backup.py:97`
  getattr guards are for loosely-mocked ctx in tests but undocumented in the function. A comment prevents future contributors from removing these.

- **[nit]** plan_callback typed as Any at `src/apm_cli/install/context.py:82`
  Known signature (plan: UpdatePlan) -> bool warrants `Callable[[UpdatePlan], bool] | None`.

#### CLI Logging Expert

- **[recommended]** backup_before_overwrite is completely silent at `src/apm_cli/install/phases/update_backup.py:67`
  Removed _purge_cached_semver_paths_for_update had a verbose_detail line. New backup emits nothing. Function returns True when staged so the caller can log -- but no caller uses the return value.
  *Suggested:* `if backup_before_overwrite(ctx, dep_ref, install_path): logger.verbose_detail(f'[*] --update: staged rollback backup for {dep_ref.get_unique_key()}')`

- **[recommended]** restore_update_backups emits no user-facing output when rolling back a declined/aborted update at `src/apm_cli/install/phases/update_backup.py:129`
  User declined the update but sees no confirmation rollback happened. Consider returning count of restored deps and emitting '[i] Restored N package(s) to pre-update state'.

- **[nit]** backup_before_overwrite docstring promises 'so the caller can log accordingly' but no caller uses the return value at `src/apm_cli/install/phases/update_backup.py:102`

#### DevX UX Expert

- **[nit]** No user-visible diagnostic when backup staging raises at `src/apm_cli/install/phases/update_backup.py:114`
  Raw OSError propagates without an actionable message. Consider wrapping with catch-and-reraise adding a one-line human hint.

- **[nit]** Duplicated predicate between _should_force_recheck and _force_semver_resolve at `src/apm_cli/deps/apm_resolver.py:886`

#### Supply Chain Security Expert

- **[recommended]** backup_path is not validated with ensure_path_within; deletions use _rrm instead of safe_rmtree at `src/apm_cli/install/phases/update_backup.py:113`
  Codebase security model requires path containment through ensure_path_within / safe_rmtree. Ad-hoc regex sanitization is a documented bug pattern. Zero-behavioral-change fix.
  *Suggested:* `from apm_cli.utils.path_security import ensure_path_within; ensure_path_within(backup_path, backup_root)` after computing backup_path. Use safe_rmtree in restore_update_backups.

- **[nit]** suppress(Exception) in restore_update_backups silences restore failures at `src/apm_cli/install/phases/update_backup.py:178`
  If restore fails, user silently ends up with missing package while lockfile references old version. Log a warning on suppressed exceptions.

#### OSS Growth Hacker

- **[nit]** CHANGELOG entry is maintainer-facing, not user-facing at `CHANGELOG.md`
  'download_callback only ran for a dependency whose install path didn't already exist' is implementation detail. User-facing rewrite: 'apm update now picks up newly published versions of transitive dependencies constrained by semver ranges (e.g. ^1.0.0)'.

- **[nit]** Two related entries (#2050 + #2053) could share a single story beat for stronger release impact.

- **[nit]** Story angle worth surfacing at release time: 'apm update now re-evaluates semver ranges at every depth -- not just direct deps.'

#### Auth Expert

No findings.

#### Doc Writer

- **[recommended]** Bug 3 (fail-closed on backup-staging failure) is absent from CHANGELOG at `CHANGELOG.md`
  Commit 813d1fbc introduces a distinct user-visible safety guarantee with no CHANGELOG bullet.
  *Suggested:* Add: '`apm update` now fails closed when staging a rollback point fails before overwriting an existing dependency: a rename or mkdir error during backup staging aborts the whole update cleanly instead of silently losing original content. (by @nadav-y) (#2053)'

- **[recommended]** docs/reference/cli/update.md does not mention transitive semver ranges are now re-evaluated at `docs/src/content/docs/reference/cli/update.md:91`
  Docs scopes re-resolution to direct deps only. Users with deep dep trees may see more packages updated than before.

- **[nit]** #2050 CHANGELOG entry says 'staged and rolled back' -- mechanism is backup-and-restore, not download staging.

- **[nit]** Attribution format deviates from predominant 0.24.0 style: use '(by @nadav-y, #2053)' not '(by @nadav-y) (#2053)'.

#### Test Coverage Expert

- **[recommended]** Plan callback decline path does not assert restore_update_backups is called with keep_new=False at `tests/unit/test_install_pipeline_orchestration.py:204`
  test_plan_callback_false_returns_early has no guardrail on the critical safety net for 'say no to update'.
  *Proof (missing):* `tests/unit/test_install_pipeline_orchestration.py::TestRunInstallPipelinePlanCallback::test_plan_callback_false_returns_early_asserts_restore` -- proves: A declined apm update restores backed-up install paths [devx, secure-by-default]

- **[recommended]** No integration-tier test for end-to-end transitive semver re-resolution backup/restore cycle at `tests/integration/test_update_e2e.py`
  Cross-module contract covered only by unit mocks. Install pipeline floor tier is integration-with-fixtures.
  *Proof (missing):* `tests/integration/test_update_e2e.py::TestUpdateE2E::test_update_dry_run_transitive_semver_restores_modules` [devx, secure-by-default]

- **[nit]** test_plan_callback_true_continues does not assert restore_update_backups(ctx, keep_new=True) is called at `tests/unit/test_install_pipeline_orchestration.py:244`

#### Performance Expert

- **[recommended]** Added network cost is O(unique_repos) RTTs via ref_resolver_cache sharing -- acceptable for explicit apm update.
  get_shared_ref_resolver reuses one RefResolver per (host, token) pair. For 5-15 transitive deps across 3-4 repos: 0-4 extra ls-remote calls. Acceptable.

- **[nit]** _should_force_recheck is zero-cost on apm install (update_refs=False short-circuits).
- **[nit]** Duplicate predicate is sync-maintenance risk, not a perf bug.
- **[nit]** backup_before_overwrite adds ~3ms I/O for 30 transitive deps -- negligible.
- **[nit]** Dedup gate prevents O(n^2) re-checks in diamond deps. Correct.

</details>

<sub>This panel is advisory. It does not block merge. Re-apply the `panel-review` label after addressing feedback to re-run.</sub>

**Comment by sergio-sisternes-epam** [comment]
Panel review complete -- this PR is approved in principle but is stacked on #2050 which needs two fixes first (failing tests + suppress(Exception) narrowing). Will approve and queue for merge once #2050 lands.

**Comment by danielmeppiel** [comment]
## APM Review Panel: `ship_now`

> Transitive semver re-resolution fix is converged, CI green, all blocking items folded -- ship it

cc @nadav-y @danielmeppiel @sergio-sisternes-epam -- a fresh advisory pass is ready for your review.

This PR addresses a correctness defect that would eventually erode community trust in apm update: transitive semver ranges were silently ignored, meaning users who expected deep-tree resolution (the behavior every mainstream package manager provides) got stale dependencies without warning. The fix is architecturally sound -- widening the existing-path gate in `_try_load_dependency_package` and threading `update_refs` through the resolver -- and avoids the over-engineering trap of rebuilding the entire resolution engine. The inline backup-before-overwrite hardening (Bug 3) is a welcome side-effect: raising on staging failure instead of swallowing it closes a data-loss window that no user would notice until it destroyed their working tree.

The shepherd-driver process has converged all ten blocking and recommended items. Test coverage is substantial: 659+ net new test lines across three test modules, plus a 12-case parametric predicate parity suite. The CHANGELOG entry is user-facing and clear, the docs/reference/cli/update.md page now describes transitive behavior explicitly, and the architecture invariant test budgets pipeline.py at its new LOC. CI is fully green across lint, two build shards, CodeQL, spec conformance, and NOTICE drift.

The three deferred items are well-scoped: integration tests for the 3-level chain and decline path require infrastructure that does not exist yet (tracked), and the apm doctor orphaned-backup check is unambiguously a new feature. None are regressions; none represent missing safety coverage on the changed code paths. The unit tests already exercise the recheck predicate across all ref kinds (semver, literal, local, artifactory) and both update_refs states.

Strategically, this fix closes a gap that incumbents handle natively. Every package manager user expects deep-tree resolution; its absence was a legitimate criticism surface. Landing this before the next minor release removes a potential adoption blocker for teams evaluating APM against npm/pip/cargo transitive semantics.

**Aligned with:** secure-by-default (backup_before_overwrite raises on staging failure, preventing silent data loss), portability-by-manifest (transitive semver ranges declared in intermediate manifests honored at any depth), governed-by-policy (update plan gate correctly reflects resolved tree before committing).

### Panel summary

| Persona | B | R | N | Takeaway |
|---|---|---|---|---|
| Python Architect | 0 | 0 | 0 | Predicate-parity parametric test confirmed; all prior architectural concerns resolved; CI green -- ship it. |
| CLI Logging Expert | 0 | 0 | 0 | Verbose breadcrumbs added; observability gap closed. |
| DevX UX Expert | 0 | 0 | 0 | Declined-plan disposition fixed; update flow now correct end-to-end. |
| Supply Chain Security | 0 | 0 | 0 | Path security guards added; no blocking vulnerabilities. |
| OSS Growth Hacker | 0 | 0 | 0 | CHANGELOG rewritten; high-trust fix -- strong release narrative beat. |
| Doc Writer | 0 | 0 | 0 | reference/cli/update.md updated; transitive re-resolution documented per spec req-rs-011. |
| Test Coverage Expert | 0 | 0 | 0 | Blocking finding resolved: patch targets corrected, all test classes pass in isolation. |
| Performance Expert | 0 | 0 | 0 | No O(n^2) regression; widened gate bounded by dedup set; backup rename cheaper than old rmtree. |

> B = blocking-severity findings, R = recommended, N = nits.
> Counts are signal strength, not gates. The maintainer ships.

### Top 3 follow-ups (deferred -- scope boundary crossed)

1. **[Integration Test]** Integration test: 3-level transitive semver chain end-to-end against live registry -- Unit coverage solid; automating the live repro prevents regression across future resolver refactors.
2. **[Integration Test]** Integration test: decline path asserts apm_modules/ untouched (not just lockfile) -- The rollback safety guarantee is load-bearing for user trust; proven by unit test + manual repro but not automated integration.
3. **[New Feature]** `apm doctor` check for orphaned `.apm-update-backup/` directories -- Defensive UX for edge cases where a crash leaves staging artifacts; not blocking but improves operational confidence.

---

**Shepherd run summary (iteration 1):**

- Copilot rounds: 1 (zero inline comments, drained after 1 fetch)
- Folded items: 10 (test fix, CANCELLED disposition, verbose logging, path_security guards, restore logging, predicate parity test, CHANGELOG rewrite, docs update, LOC budget, staging rollback fix)
- Deferred items: 3 (integration tests, apm doctor) -- scope boundary: integration test infrastructure and new feature
- CI iterations: 2 (first push: staging regression + LOC budget failure; second push: all green)
- Lint evidence: ruff check + ruff format + pylint R0801 + auth-signals all silent on final SHA
- CI evidence: https://github.com/microsoft/apm/actions/runs/29182970554 (all SUCCESS)
- Head SHA: 5706407225c6af8feb3cf0fae8659c79af0e4edf
- Mergeable: MERGEABLE | MergeStateStatus: BLOCKED (pending required review -- normal gate)

<details>
<summary>Copilot signals reviewed (round 1)</summary>

Copilot posted 1 overview review (no inline comments).
- Review ID 4636176721: OVERVIEW -- "Copilot reviewed 9 out of 9 changed files... generated 1 comment." -- Classification: NOT-LEGIT for inline purposes (no actionable inline finding; the single generated comment is a summary, not a defect report). No inline items to fold.

</details>

**Comment by nadav-y** [comment]
Putting on draft to rebase after https://github.com/microsoft/apm/pull/2052 /  https://github.com/microsoft/apm/pull/2155 merge and look into comments and tests. 
I'll reopen soon

-- edit: Done
